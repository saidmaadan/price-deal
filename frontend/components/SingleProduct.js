import React, { Component } from 'react';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import Error from './ErrorMessage';
import styled from 'styled-components';
import Head from 'next/head';


const SingleProductStyles = styled.div`
  max-width: 1200px;
  margin: 2rem auto;
  box-shadow: ${props => props.theme.bs};
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  min-height: 800px;
  img {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }
  .details {
    margin: 3rem;
    font-size: 2rem;
  }
`;

const SINGLE_PRODUCT_QUERY = gql`
  query SINGLE_PRODUCT_QUERY($id: ID!) {
    product(where: { id: $id }) {
      id
      title
      description
      largeImage
    }
  }
`;
class SingleProduct extends Component {
  render() {
    return (
      <Query
        query={SINGLE_PRODUCT_QUERY}
        variables={{
          id: this.props.id,
        }}
      >
        {({ error, loading, data }) => {
          if (error) return <Error error={error} />;
          if (loading) return <p>Loading...</p>;
          //if (!data.product) return <p>No Product Found for {this.props.id}</p>;
          const product = data.product;
          return (
            <SingleProductStyles>
              <Head>
                <title>Price Deal | {product.title}</title>
              </Head>
              <img src={product.largeImage} alt={product.title} />
              <div className="details">
                <h2>Viewing {product.title}</h2>
                <p>{product.description}</p>
              </div>
            </SingleProductStyles>
          );
        }}
      </Query>
    );
  }
}

export default SingleProduct;
export { SINGLE_PRODUCT_QUERY };
