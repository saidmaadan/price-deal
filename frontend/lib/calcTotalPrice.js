export default function calcTotalPrice(cart) {
  return cart.reduce((tally, cartProduct) => {
    if (!cartProduct.product) return tally;
    return tally + cartProduct.quantity * cartProduct.product.price;
  }, 0);
}
