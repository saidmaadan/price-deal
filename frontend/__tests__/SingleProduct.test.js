import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import SingleProduct, { SINGLE_PRODUCT_QUERY } from '../components/SingleProduct';
import { MockedProvider } from 'react-apollo/test-utils';
import { fakeProduct } from '../lib/testUtils';

describe('<SingleProduct/>', () => {
  it('renders with proper data', async () => {
    const mocks = [
      {
        // when someone makes a request with this query and variable combo
        request: { query: SINGLE_PRODUCT_QUERY, variables: { id: '123' } },
        // return this fake data (mocked data)
        result: {
          data: {
            product: fakeProduct(),
          },
        },
      },
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleProduct id="123" />
      </MockedProvider>
    );
    expect(wrapper.text()).toContain('Loading...');
    await wait();
    wrapper.update();
    // console.log(wrapper.debug());
    expect(toJSON(wrapper.find('h2'))).toMatchSnapshot();
    expect(toJSON(wrapper.find('img'))).toMatchSnapshot();
    expect(toJSON(wrapper.find('p'))).toMatchSnapshot();
  });

  it('Errors with a not found Product', async () => {
    const mocks = [
      {
        request: { query: SINGLE_PRODUCT_QUERY, variables: { id: '123' } },
        result: {
          errors: [{ message: 'Products Not Found!' }],
        },
      },
    ];
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleProduct id="123" />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    console.log(wrapper.debug());
    const product = wrapper.find('[data-test="graphql-error"]');
    expect(product.text()).toContain('Products Not Found!');
    expect(toJSON(product)).toMatchSnapshot();
  });
});
