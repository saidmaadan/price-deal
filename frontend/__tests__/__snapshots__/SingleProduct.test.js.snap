// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleProduct/> Errors with a not found Product 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Shoot!
  </strong>
  Products Not Found!
</p>
`;

exports[`<SingleProduct/> renders with proper data 1`] = `
<h2>
  Viewing 
  dogs are best
</h2>
`;

exports[`<SingleProduct/> renders with proper data 2`] = `
<img
  alt="dogs are best"
  src="dog.jpg"
/>
`;

exports[`<SingleProduct/> renders with proper data 3`] = `
<p>
  dogs
</p>
`;
